### DISCUSSION TITLE
Using larger runners - Compile curl faster on github

### DISCUSSION THREAD

**Discussion by pheiduck** [curl/curl#10600] [Ideas]
@bagder
Maybe we could implement this. This feature is only available for organizations.
https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners

**Comment by bagder** [comment]
Why?

**Reply by andy5995** [reply]
Because, with new larger runners, teams can spend less time managing infrastructures and more time shipping code using fast, scalable, secure compute resources fully managed by GitHub.

At least, that's what's written on [the sign-up page](https://github.com/features/github-hosted-runners/signup). xD Reading the link @pheiduck provided, it looks like you could build with up to 64 cores in the CI, instead of the current two.

**Reply by pheiduck** [reply]
> Why?

On the other hand, I heard that you have a new DEV system. I congratulate you Daniel Stenberg for being the 2nd best developer, you deserve it 😉🎉

**Comment by andy5995** [comment]
> @bagder Maybe we could implement this. This feature is only available for organizations. https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners

@pheiduck Nice find. ;) I hadn't heard about this... What other secrets do you possess?

**Reply by gvanem** [reply]
**andy5995**> could build with up to 64 cores in the CI, instead of the current two.

With which compiler do you mean? With the MSVC compiler I compile libcurl here locally in 10 sec with `cl -MP ..`. Uses all my 24 cores.
So could `msbuild` do on a Github Workflow AFAICS.

**Reply by andy5995** [reply]
@gvanem Good question. All the info available is in the docs. I don't know any more than that. :)

**Comment by bagder** [comment]
Well yes, obviously using more cores would build curl a little faster. But since we want all CI jobs to complete as fast as possible, just getting a few to run much faster won't help much. Especially when those are the jobs that already now complete among the first ones.

Also: the most time consuming parts of the CI jobs are not the actual build, it is the running of the test suite. In particular when valgrind is enabled. That is still using a single core no matter how many cores that are available. A more worthwhile work would probably be to introduce parallel support for the test suite.

**Reply by andy5995** [reply]
I didn't see a ticket for that. What needs to happen to make the tests run in parallel? I thought they already did because of the results from [this PR](https://github.com/curl/curl/pull/10232#issuecomment-1372165662). Is there something in the automake that causes them to ignore the `-j` argument if it's higher than 1?

**Comment by dfandrich** [comment]
I looked into parallelizing the tests about 6 years ago and had a hacked-up
branch that would run tests from different protocols in parallel. But, because
most tests use HTTP, that amount of parallelism didn't reduce the test time
enough to make it worth cleaning up and submitting—not even 50%.

The problem with more parallelization is that the test suite is designed to run
only a single protocol server at once; that assumption is baked into many
places and would be a big job to overcome. But, if a test run can be brought
down from 500 seconds to 10 seconds, which is probably in the realm of
possibility, that total savings in man-hours over the remaining lifetime of
curl would make it worthwhile.

**Reply by bagder** [reply]
> The problem with more parallelization is that the test suite is designed to run only a single protocol server at once; that assumption is baked into many places and would be a big job to overcome

Yeah, I think I also would qualify that as a "big job", but I still consider that a worthy project to explore. If we could fire up 10 HTTP servers and use those instances properly, the HTTP tests could finish 10 times faster. And we don't have to stop at 10...
